dbms/src/Flash/Pipeline/Pipeline.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Flash/Executor/ResultHandler.h>
+#include <Flash/Pipeline/Exec/PipelineExec.h>
+
+#include <deque>
+#include <vector>
+
+namespace tipb
+{
+class DAGRequest;
+}
+
+namespace DB
+{
+class Pipeline;
+using PipelinePtr = std::shared_ptr<Pipeline>;
+using Pipelines = std::vector<PipelinePtr>;
+
+class Event;
+using EventPtr = std::shared_ptr<Event>;
+using Events = std::vector<EventPtr>;
+
+class PhysicalPlanNode;
+using PhysicalPlanNodePtr = std::shared_ptr<PhysicalPlanNode>;
+
+class PipelineExecutorStatus;
+
+class Pipeline : public std::enable_shared_from_this<Pipeline>
+{
+public:
+    explicit Pipeline(UInt32 id_)
+        : id(id_)
+    {}
+
+    void addPlanNode(const PhysicalPlanNodePtr & plan_node);
+
+    void addChild(const PipelinePtr & child);
+
+    void toTreeString(FmtBuffer & buffer, size_t level = 0) const;
+
+    // only used for test to get the result blocks.
+    void addGetResultSink(ResultHandler result_handler);
+
+    PipelineExecGroup toExecGroup(PipelineExecutorStatus & exec_status, Context & context, size_t concurrency);
+
+    Events toEvents(PipelineExecutorStatus & status, Context & context, size_t concurrency);
```

Comment:
So you're saying that the `PipelineEvent` holds the `PhysicalPlanNode` directly and calls `PhysicalPlanNode.toPipelineExec` directly?

Reply:
This is because the physical plan node is currently held by `Pipeline` and the physical plan node is needed to create the `PipelineExec`.
Maybe it would be better to rename `toExecGroup` to `buildExecGroup`?